crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,770 @@
+        .collect()
+}
+
+#[derive(Clone, Debug)]
+#[pyclass]
+pub struct CommutationLibrary {
+    pub library: Option<HashMap<(String, String), CommutationLibraryEntry>>,
+}
+
+impl CommutationLibrary {
+    fn check_commutation_entries(
+        &self,
+        first_op: &OperationRef,
+        first_qargs: &[Qubit],
+        second_op: &OperationRef,
+        second_qargs: &[Qubit],
+    ) -> Option<bool> {
+        if let Some(library) = &self.library {
+            match library.get(&(first_op.name().to_string(), second_op.name().to_string())) {
+                Some(CommutationLibraryEntry::Commutes(b)) => Some(*b),
+                Some(CommutationLibraryEntry::QubitMapping(qm)) => qm
+                    .get(&get_relative_placement(first_qargs, second_qargs))
+                    .copied(),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }
+}
+
+#[pymethods]
+impl CommutationLibrary {
+    #[new]
+    fn new(py_any: Option<Bound<PyAny>>) -> Self {
+        match py_any {
+            Some(pyob) => CommutationLibrary {
+                library: pyob
+                    .extract::<Option<HashMap<(String, String), CommutationLibraryEntry>>>()
+                    .unwrap(),
+            },
+            None => CommutationLibrary {
+                library: Some(HashMap::new()),
+            },
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum CommutationLibraryEntry {
+    Commutes(bool),
+    QubitMapping(HashMap<SmallVec<[Option<Qubit>; 2]>, bool>),
+}
+
+impl<'py> FromPyObject<'py> for CommutationLibraryEntry {
+    fn extract_bound(b: &Bound<'py, PyAny>) -> Result<Self, PyErr> {
+        if let Ok(b) = b.extract::<bool>() {
+            return Ok(CommutationLibraryEntry::Commutes(b));
+        }
+        let dict = b.downcast::<PyDict>()?;
+        let mut ret = hashbrown::HashMap::with_capacity(dict.len());
+        for (k, v) in dict {
+            let raw_key: SmallVec<[Option<u32>; 2]> = k.extract()?;
+            let v: bool = v.extract()?;
+            let key = raw_key.into_iter().map(|key| key.map(Qubit)).collect();
+            ret.insert(key, v);
+        }
+        Ok(CommutationLibraryEntry::QubitMapping(ret))
+    }
+}
+
+impl ToPyObject for CommutationLibraryEntry {
+    fn to_object(&self, py: Python) -> PyObject {
+        match self {
+            CommutationLibraryEntry::Commutes(b) => b.into_py(py),
+            CommutationLibraryEntry::QubitMapping(qm) => {
+                let out_dict = PyDict::new_bound(py);
+
+                qm.iter().for_each(|(k, v)| {
+                    out_dict
+                        .set_item(
+                            PyTuple::new_bound(py, k.iter().map(|q| q.map(|t| t.0))),
+                            PyBool::new_bound(py, *v),
+                        )
+                        .ok()
+                        .unwrap()
+                });
+                out_dict.unbind().into_any()
+            }
+        }
+    }
```

Comment:
To build a `PyDict` more safely without rust allocations, you could do it directly from an iterator using this, which would avoid you having to unwrap and reduce the risk of `panic!`.

```suggestion
    fn to_object(&self, py: Python) -> PyObject {
        match self {
            CommutationLibraryEntry::Commutes(b) => b.into_py(py),
            CommutationLibraryEntry::QubitMapping(qm) => {
                qm.iter().map(|(k, v)| (
                            PyTuple::new_bound(py, k.iter().map(|q| q.map(|t| t.0))),
                            PyBool::new_bound(py, *v),
                        )
                ).into_py_dict_bound(py).unbind().into()
            }
        }
    }
```

Reply:
Done in 12d0a05